modules/tinymce/src/themes/silver/main/ts/ui/core/ChoiceControls.ts

```diff
@@ -0,0 +1,150 @@
+        lastApi.clear,
+        // If we do have a menu item for the current state, highlight it
+        (api) => {
+          lastApi.set({
+            destroy: () => api.setActive(false)
+          });
+          api.setActive(true);
+        }
+      );
+    };
+
+    editor.on('NodeChange', callback);
+
+    return Arr.map(
+      options,
+      (value, i) => ({
+        type: 'togglemenuitem',
+        text: spec.display(value),
+        onSetup: (api) => {
+          apis[spec.normalise(value)] = api;
+
+          if (i + 1 === options.length) {
+            // run the callback once on startup (on the last option so that we know the apis map has been set up)
+            callback();
+          }
+
+          return () => {
+            // only clean up global things once
+            if (i === 0) {
+              editor.off('NodeChange', callback);
+              lastApi.clear();
+            }
+          };
+        },
+        onAction: () => spec.setCurrent(editor, value)
+      })
+    );
+  };
+
+  editor.ui.registry.addMenuButton(spec.name, {
+    tooltip: spec.text,
+    icon: spec.icon,
+    fetch: (callback) => callback(getMenuItems())
+  });
+
+  editor.ui.registry.addNestedMenuItem(spec.name, {
+    type: 'nestedmenuitem',
+    text: spec.text,
+    getSubmenuItems: getMenuItems
+  });
+};
+
+const lineHeightSpec: ControlSpec<string> = {
+  name: 'lineheight',
+  text: 'Line height',
+  icon: 'line-height',
+
+  getOptions: Settings.getLineHeightFormats,
+  normalise: (input) => Dimension.normalise(input, [ 'fixed', 'relative', 'empty' ]).getOr(input),
+  display: Fun.identity,
+
+  getCurrent: (editor) => Optional.some(editor.queryCommandValue('LineHeight')),
+  setCurrent: (editor, value) => editor.execCommand('LineHeight', false, value)
+};
+
+const languageSpec: ControlSpec<ContentLanguage> = {
+  name: 'language',
+  text: 'Language',
+  icon: 'translate',
+
+  getOptions: Settings.getContentLangugaes,
+  normalise: (input) => Type.isUndefined(input.customCode) ? input.code : `${input.code}/${input.customCode}`,
+  display: (input) => input.title,
+
+  getCurrent: (editor) => {
+    const selection = SugarElement.fromDom(editor.selection.getRng().startContainer);
+    return TransformFind.ancestor(selection, (node) => {
+      if (!SugarNode.isElement(node)) {
+        return Optional.none();
+      }
+
+      const codeOpt = Attribute.getOpt(node, 'lang');
+      const customCode = Attribute.getOpt(node, 'data-mce-lang').getOrUndefined();
+      // The actual language title doesn't matter, because all we're going to do
+      // is pass this to normalise (which ignores the title)
+      const title = '';
+
+      return codeOpt.map((code): ContentLanguage => ({ code, customCode, title }));
+    });
+  },
+  setCurrent: (editor, lang) => {
+    editor.formatter.toggle('lang', {
+      value: lang.code,
+      customValue: lang.customCode
+    });
+    editor.nodeChanged();
```

Comment:
This needs to be in an undomanager transaction, otherwise you won't be able to undo the changes.

Reply:
Maybe this should be a command I think most other UI things can be done though commands. Like a `mceToggleLanguage` or just `ToggleLanguage` if we are getting rid of the `mce` prefixes?

Reply:
Added it to a transaction for now, will have a chat among the team about whether or not to add a command.

Reply:
Now a command, currently just called `Lang` for parity with the other commands that are just wrappers around a format.